### ISSUE TITLE
Fully support prepared statements with pgbouncer 1.21.0

### ISSUE THREAD

**Issue by jineshshah36** [prisma/prisma#21531]
## Problem

Today, the way prepared statements are supported with pgbouncer is suboptimal. Prisma will clear all prepared statements in the connection to avoid mismatches. Now that pgbouncer supports prepared statements as of 1.21.0, It would be great to understand if we need to append `?pgbouncer=true` or if we can simply run prisma without being in "pgbouncer" mode.

## Suggested solution

Clarity on how to fully take advantage of prepared statements with pgbouncer 1.21.0

**Comment by MorenoMdz** [comment]
I would assume Prisma needs to be updated to support this, based on the information [here](https://github.com/prisma/prisma/discussions/21372#discussioncomment-7306342).

Would be great to have this prioritized, as currently we are facing some issues because of the extra latency.

**Comment by janpio** [comment]
Did anyone try running their test suites against a this new PgBouncer version either with or without `pgbouncer=true`? That would probably be a good first data point.

**Comment by MorenoMdz** [comment]
> Did anyone try running their test suites against a this new PgBouncer version and `max_prepared_statements` configured, either with or without `pgbouncer=true`? That would probably be a good first data point.

I will try to test it this way and report back

**Comment by MorenoMdz** [comment]
> Did anyone try running their test suites against a this new PgBouncer version and `max_prepared_statements` configured, either with or without `pgbouncer=true`? That would probably be a good first data point.

So I tested with this setup:

1. PG running locally withing docker + pgbouncer 1.21 running natively locally
* Without the pgbouncer param, it does throw
* With the pgbouncer param, it works fine

2. PG running from cloud, using the direct access url, connected thru the same pgbouncer 1.21 running natively, same results.

**Comment by janpio** [comment]
We actually have reports now that _with_ `pgbouncer=true` you will get an error: https://github.com/prisma/prisma/issues/21635
That conflicts with your observations @MorenoMdz, correct?

What error message exactly are you getting in your testing?

**Comment by MorenoMdz** [comment]
> We actually have reports now that _with_ `pgbouncer=true` you will get an error: #21635 That conflicts with your observations @MorenoMdz, correct?
> 
> What error message exactly are you getting in your testing?

Can it be because of the Prisma version? We are using `5.4.2` but I would be surprised if it was only for my case. Here if I connect to my local PG or cloud PG without `pgbouncer=true` thru my local instance of pgbouncer, it will error.

This is the pgbouncer config I used, I don't know if this would have any affect.

```ini
[databases]
db = host=xxx dbname=xxx user=xxx password=xxx


[pgbouncer]
listen_addr = 127.0.0.1
listen_port = 6432
auth_type = plain
auth_file = pgbouncer_users.txt
logfile = pgbouncer.log
pidfile = pgbouncer.pid
ignore_startup_parameters = extra_float_digits

```

**Comment by janpio** [comment]
How is your pgbouncer 1.21 configured? The new functionality is only enabled when `max_prepared_statements` is set to a non `0` value. Otherwise 1.21 behaves exactly like older PgBouncer versions (which would match _your_ observations).

**Comment by MorenoMdz** [comment]
> That might be the problem: The new functionality is only enabled when `max_prepared_statements` is set to a non `0` value. Otherwise 1.21 behaves exactly like older PgBouncer versions (which would match _your_ observations).

That must be it them, I assumed incorrectly it was supported by default, going to try again.

**Comment by janpio** [comment]
Unfortunately not. I have not looked deeper into this.

**Comment by MorenoMdz** [comment]
> Unfortunately not. I have not looked deeper into this.

Ok good news:

```ini
[databases]
db = host=xxx dbname=xxx user=xxx password=xxx


[pgbouncer]
listen_addr = 127.0.0.1
listen_port = 6432
auth_type = plain
auth_file = pgbouncer_users.txt
logfile = pgbouncer.log
pidfile = pgbouncer.pid
ignore_startup_parameters = extra_float_digits
; max_prepared_statements = 1000 
```

With this setup:
* Pgbouncer running on my machine, targeting PG in the cloud (Aiven) using the direct access URL
* Running Prisma 5.4.2 in Nest, calling the requests from Swagger
* No `pgbouncer=true` param in the URL
* Got consistent 400ms in a specific endpoint that before was always above 1000ms


(Keep in mind I am in LATAM right now and the server is in central-us so the extra latency is expected)

Before
![image](https://github.com/prisma/prisma/assets/29472480/e88389f1-c87d-4d14-a29b-cd670f86b49e)

After
![image](https://github.com/prisma/prisma/assets/29472480/b7a42da7-c4b5-4292-9006-1d6ea346b8e0)

I can run an artillery session from US to get a better load test if you want.

What was causing it to break first is that I had the pool URL of my cloud DB, changing it to the direct URL made it work. What I am not sure yet is why it was breaking against my local PG.

Note that having `max_prepared_statements = 1000` or not didn't made any difference in this small test.

**Comment by jineshshah36** [comment]
We also tested `1.21.0` with `max_prepared_statements = 1000` and removing `pgbouncer=true` on prisma `5.4.2` and everything *just works*! We're seeing improved query latency as well and no issues. I think it would be a good idea to deprecate `pgbouncer=true` and/or recommend using pgbouncer `1.21.0` with prepared statements enabled. This one change has improved our query latency more than any change prisma has shipped (including all of the latest perf improvements). I wouldn't sleep on this.

**Comment by janpio** [comment]
Can you also test running Introspection and/or Migration via the PgBouncer you setup? Thanks @jineshshah36.
(We are now also starting to test this, but any help is welcome in figuring out the details.)

**Comment by jineshshah36** [comment]
Our setup doesn't allow for that very easily. We have a separate proxy for that stuff that does not use pgbouncer.

**Comment by janpio** [comment]
Hm, after reading https://www.pgbouncer.org/config.html#max_prepared_statements in detail I would actually be surprised if this _just works_ without `pgbouncer=true`. Multiple instances of Prisma Client would probably try to prepare different queries with the same name for the prepared statement, leading to an error message when actually executing the query. We'll have to get a reproduction for this scenario.

**Comment by janpio** [comment]
Seems that part is probably fine as PgBouncer deals with that properly.
A problem we found already is when you use `schema=foo` in your connection string, that leads to this error as PgBoucner fully does not support `search_path` still:
```
PrismaClientInitializationError: Error querying the database: FATAL: unsupported startup parameter: search_path
```

**Comment by jineshshah36** [comment]
I'm not sure I follow. I'm not aware of this `schema` argument. We have one schema (`public`) and all works ok

**Comment by JelteF** [comment]
Isn't the search_path problem an issue with pgbouncer=true as well? In any case it seems rather unrelated to prepared statements.

BTW, it is supported if you use PgBouncer together with Citus: https://www.pgbouncer.org/config.html#track_extra_parameters

**Comment by janpio** [comment]
Yes @JelteF, but it means that upgrading to PgBouncer 1.21.0 does not magically make `pgbouncer=true` (for our own PgBouncer mode) obsolete, as without it you will directly run into `FATAL: unsupported startup parameter: search_path` if you use a non standard "schema" (which with Prisma you configure in the connection string via `?schema=foo`, that is then translated into a `SET search_path` query under the hood).

Seems @jineshshah36 is lucky enough that he is not using that. Many other users of Prisma are, so we need to find a solution there. (Which is very possible, Prisma fully qualifies most queries so the search_path is mostly used for raw queries and some other things that one probably would not want to do via PgBouncer anyway)

**Comment by JelteF** [comment]
I guess part of my question was not clear: how does pgbouncer=true make ?schema=foo work? Does it start qualifying every table? 

Also working around the search_path issue is fairly simple for an operator. They can either change the default search_path for the relevant postgres user. Or they can use connect_query to set the search_path: connect_query=set search_path=foo;

**Comment by janpio** [comment]
Ah, understood: `pgbouncer=true` makes Prisma skip the `SET search_path = ...` query basically only. We qualify tables by default - except for user-provided raw queries, which might not work as expected in PgBouncer mode.

Unfortunately, we have to assume that PgBouncer is out of control of most of our users as they just get a pre-configured instance by DigitalOcean, Supabase or Neon. We just need to make it work as well as possible. (Which is now pretty amazing, with the named prepared statement support - just need to cover it with tests now and make a call on how `search_path` will be handled)

**Comment by JelteF** [comment]
> Ah, understood: `pgbouncer=true` makes Prisma skip the `SET search_path = ...` query basically only. We qualify tables by default - except for user-provided raw queries, which might not work as expected in PgBouncer mode.

Ah okay, that makes sense. But if I understand correctly, then using `pgbouncer=true` in the connection string completely undoes the effect of the `schema=foo` argument in the same connection string. If that's indeed the case then it seems the easiest to simply say that people should not use the `schema=foo` argument in the connection string when connecting to pgbouncer 1.21.0. So where the connection string would first be:
```
postgresql://some-host-name.com?pgbouncer=true&schema=foo
```
It would now simply be:
```
postgresql://some-host-name.com
```

**Comment by MorenoMdz** [comment]
One thing to notice, upgrading your PG bouncer can be very tricky depending on your deployment pipeline. Even in the scenario that all your Prisma queries would work with pgbouncer 1.21 without the `pgbouncer=true` param, as soon as you upgrade your pooler to 1.21 your Prisma queries will start failing. On the other hand, if you remove `pgbouncer=true` first then upgrade your DB, your queries will also start failing. 

In order to not have downtime and or have to do a full server swap deploy, ideally Prisma would support the new version of pgbouncer with or without `pgbouncer=true`, is it possible at all?

cc @janpio

**Comment by MorenoMdz** [comment]
One extra data point. Our cloud provider (Aiven) recently release support for the latest version of pgbouncer, I just tested for the last hour and the results are mixed. 
* Running Prisma without `pgbouncer=true` is indeed faster overall. 
* running with the pgbouncer param will cause all queries I tested to error.
* Most of our queries work, including some prisma RAW queries, but a lot of queries will now fail with errors that still mention pgbouncer, for example:


```
Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(PostgresError { code: "26000", message: "prepared statement \"PGBOUNCER_10\" does not exist", severity: "ERROR", detail: None, column: None, hint: None }), transient: false })
```

I triple-checked that our internal backend server is not using `pgbouncer` in the connection URL.

**Comment by JelteF** [comment]
So, that error definitely means somehow pgbouncer thinks a query is prepared on the server connection, but somehow it isn't. Are you sure there are no clients at all using `pgbouncer=true`? Because if there's only one client with that enabled that setting it can "poison" all the server connections so that other clients start failing.

**Comment by MorenoMdz** [comment]
> So, that error definitely means somehow pgbouncer thinks a query is prepared on the server connection, but somehow it isn't. Are you sure there are no clients at all using `pgbouncer=true`? Because if there's only one client with that enabled that setting it can "poison" all the server connections so that other clients start failing. 

There could be some from our dev team pinging it as it is an internal dev db I am testing on, I will also double check our cloud functions.

Does that mean we can not have the same server in use with two different type of requests?

**Comment by JelteF** [comment]
> In order to not have downtime and or have to do a full server swap deploy, ideally Prisma would support the new version of pgbouncer with or without `pgbouncer=true`, is it possible at all?

The next PgBouncer release should support pgbouncer=true too since I merged the PR to support `DEALLOCATE ALL` there, but the next official release will probably be after the new years. (It's a new feature and so far only the prisma client are impacted, so I don't think rushing it out as a a patch release is worth the risk/hassle with holidays coming up). 

For now as long as you keep max_prepared_statements at its default of 0 you can continue to use pgbouncer=true even with the new pgbouncer version (but you won't benefit from the new feature). Then when the next one releases you can try a rolling release.

**Comment by MorenoMdz** [comment]
> > In order to not have downtime and or have to do a full server swap deploy, ideally Prisma would support the new version of pgbouncer with or without `pgbouncer=true`, is it possible at all?
> 
> 
> 
> The next PgBouncer release should support pgbouncer=true too since I merged the PR to support `DEALLOCATE ALL` there, but the next official release will probably be after the new years. (It's a new feature and so far only the prisma client are impacted, so I don't think rushing it out as a a patch release is worth the risk/hassle with holidays coming up). 
> 
> 
> 
> For now as long as you keep max_prepared_statements at its default of 0 you can continue to use pgbouncer=true even with the new pgbouncer version (but you won't benefit from the new feature). Then when the next one releases you can do a rolling deployment of your clients. 

Do I need to explicitly set max_prepared_statements to 0? Because as soon as I updated to the v1.21.0 our queries started failing. I have requested to our provider to expose this setting so we can configure it.

As for the next release I do understand the concerns of testing it before the holidays, but at the same time I can assure the gains are large enough that as an end user being impacted by 200ms+ queries it feels important enough to have a patch :(

**Comment by janpio** [comment]
If you "set max_prepared_statements to 0" you disable the new named prepared statement support of PgBouncer and will need to keep using `pgbouncer=true` @MorenoMdz.

The problem of the new PgBouncer version failing with `pgbouncer=true` you describe is tracked at https://github.com/prisma/prisma/issues/21635 (which has links to the pgBouncer issues and PR) and will indeed need to be fixed before Prisma users can use the new version with the old `pgbouncer=true` configuration.

---

@JelteF From our side it would be highly desirable if the `DEALLOCATE ALL` fix could be released soon. Any cloud provider that updates to 1.21.0 will instantly break all Prisma Client application that use `pgbouncer=true` (which should be most to all, as it was necessary in the past) - see https://github.com/prisma/prisma/issues/21635 as a description what happened when Neon did that - and require an update of the application configuration to remove `pgbouncer=true`. 
We'll try to handle the error in the next Prisma release to tell users in a nice error message what is going on, but that will only apply to future releases of course and many of our users will be running older versions.

**Comment by janpio** [comment]
So to recap:

- PgBouncer < 1.21.0 => Needs `pgbouncer=true` to work correctly
- PgBouncer 1.21.0, max_prepared_statements = 0 => Needs `pgbouncer=true` to work correctly
- PgBouncer 1.21.0, max_prepared_statements > 0 => Will fail with `pgbouncer=true` (because of https://github.com/prisma/prisma/issues/21635)
- PgBouncer 1.21.0, max_prepared_statements > 0 => Should work find without `pgbouncer=true` (still confirming)

Does that match what you are seeing @MorenoMdz?

**Comment by JelteF** [comment]
> Any cloud provider that updates to 1.21.0 will instantly break all Prisma Client application that use pgbouncer=true

To be clear, that should be "any cloud provider deploys 1.21.0 and changes the `max_prepared_statements` setting from its default".

The new prepared statement support was deliberately left as "off" by default in PgBouncer for now, specifically to not break existing users in case of any bugs, perf impact or weird interactions. If your cloud providers turns the feature on by default anyway for all their existing users, then that seems like something you should talk to your cloud provider about. 

I do agree that the current breakage that has been happening for Prisma applications is not great though. But I'll be on honeymoon for a month soon, so that's why I'm quite hesitant at creating a new feature-release before that, since it could possibly introduce other bugs. Especially since the current issue described here can be considered a weird interaction, instead of a bug (i.e. current non-support for `DEALLOCATE ALL` with `max_prepared_statetements >= 0` is explicitly documented).

P.S. Prisma is the only client that I know of that uses `DEALLOCATE ALL` to work around PgBouncer issue. All others use unnamed prepared statements instead. i.e. they use the empty string (`""`) as the name for all their prepared statements, this means that they don't need to be deallocated, because a prepared statement with the empty string name can be recreated without deallocating it.

**Comment by MorenoMdz** [comment]
> So to recap:
> 
> * PgBouncer <1.21.0 => Needs `pgbouncer=true` to work correctly
> * PgBouncer 1.21.0, max_prepared_statements = 0 => Needs `pgbouncer=true` to work correctly (same behavior as <1.21.0)
> * PgBouncer 1.21.0, max_prepared_statements > 0 => Will fail with `pgbouncer=true` (because of [Prisma PgBouncer mode doesn't play nice with PGBouncer v1.21.0 #21635](https://github.com/prisma/prisma/issues/21635))
> * PgBouncer 1.21.0, max_prepared_statements > 0 => Should work fine without `pgbouncer=true` (still confirming)
> * PgBouncer >1.21.0 (future release) => Should work fine with and without `pgbouncer=true`
> 
> Does that match what you are seeing @MorenoMdz?

Yes, with a couple things to add.
* Our cloud provider has max_prepared_statements set to 100 vs the default of 0.
* Having clients with different configurations, one with `pgbouncer=true` and others without it, will cause pgbouncer to error out on _some_  of the prepared statements requests. [info](https://github.com/prisma/prisma/issues/21531#issuecomment-1793875613)

**Comment by janpio** [comment]
@JelteF Congratulations!

> All other clients use unnamed prepared statements instead. i.e. they use the empty string ("") as the name for all their prepared statements, this means that they don't need to be deallocated, because a prepared statement with the empty string name can be recreated without deallocating it.

If only this had been documented somewhere, we might have ended up with the same solution here :D

**Comment by MorenoMdz** [comment]
Would it be a too large of a lift to implement the empty string approach now?

**Comment by janpio** [comment]
Yes, and it is also not necessary anymore if we can confirm that case 4) ("4. PgBouncer 1.21.0, max_prepared_statements > 0") works fine without `pgbouncer=true` (when no other Clients still use that).

**Comment by guyromb** [comment]
I am facing the same issue. What is the recommended fix to still be able to use pgbounce pool with prisma? (pgbounce set to transactional mode). I tried to remove pgbouncer=true or keep it - but none works for me.. max_prepared_statements is 0.
I am still getting error 26000 `ERROR: prepared statement \"PGBOUNCER_36\" does not exist`

**Comment by janpio** [comment]
You posted the same in 3 issues now @guyromb. Please open a new issue and describe your full situation and problem. Thank you.

**Comment by MorenoMdz** [comment]
> I am facing the same issue. What is the recommended fix to still be able to use pgbounce pool with prisma? (pgbounce set to transactional mode). I tried to remove pgbouncer=true or keep it - but none works for me.. max_prepared_statements is 0. I am still getting error 26000 `ERROR: prepared statement \"PGBOUNCER_36\" does not exist`

Can you double check that running Prisma without `pgbouncer=true` connecting to pgbouncer v1.21.0 with `max_prepared_statements = 0` you are still getting those errors?

Be 100% sure that NO other client is also connecting at the same time with a different mode, with `pgbouncer=true` or not and vice versa. We had a lot of issues with team members connecting from their local computers with the param in the URL still causing the whole `pgbouncer` server to start throwing these errors.

**Comment by guyromb** [comment]
I created a specific issue for that https://github.com/prisma/prisma/issues/21877
max_prepared_statements=0 (That's the default in digitalocean managed postgres)
pgbouncer=true is removed in from the string of all applications.

Strangely enough it used to work before with pgbouncer=true without any issues before. The issue started a few hours ago.

**Comment by MorenoMdz** [comment]
@janpio @JelteF would this plan work for us to migrate?

----
At a low usage time, deploy everything to use the direct URL (no pgbouncer true, different port) so no pooling with be done. We can set the max connections in the DB server to a higher number temporarily and maybe also increase the server config for a couple hours.

Ask Aiven (our cloud provider) to change the max prepared statements setting to 100, a nso clients are going to be connecting to the pool, no issues would happen.

Redeploy our cloud functions to connect to the pool, without `pgbouncer=true` , check that no errors happened.

Redeploy our Node server to connect back to the pool.

Change the configuration/connections number down to what it is now.
---

**Comment by janpio** [comment]
Yes, that would probably a be a reasonable way to avoid any user facing errors or downtime.

If that is not 100% required, you can simplify a lot: Ask Aiven to increase `max_prepared_statements` to a non 0 value, then instantly redeploy all your Node servers to not use `pgbouncer=true` any more - that will probably still put PgBoucner into a bad state because before the deployment some `DEALLOCATE ALL` will be run. But then restart the PgBounce instance so it is in a clean state - and you should be fine.

**Comment by MorenoMdz** [comment]
> Yes, that would probably a be a reasonable way to avoid any user facing errors or downtime.
> 
> If that is not 100% required, you can simplify a lot: Ask Aiven to increase `max_prepared_statements` to a non 0 value, then instantly redeploy all your Node servers to not use `pgbouncer=true` any more - that will probably still put PgBoucner into a bad state because before the deployment some `DEALLOCATE ALL` will be run. But then restart the PgBounce instance so it is in a clean state - and you should be fine.

The blocker is that we have a whole replication setup happening with cloud functions, we are finalizing a migration from Firestore to PG and there are a few hundred requests happening per second still just for the live replication from FS->PG. If we change the cloud functions and the Node server to use the direct URL there is the risk of issues due to rejected requests because of saturated connections from the DB server.

One extra issue is that deploying our Nest instance can take up to 10-30min depending on our provider (Render) and the cloud functions also need a redeploy to update the URL.

That said, we might go ahead and upgrade it tomorrow night as it will likely to be a low usage time in our systems due to the holidays. I will update here depending on how it goes 👍

**Comment by janpio** [comment]
Yes it does, with that fix the `DEALLOCATE ALL` is not only executed on the database connection, removing its prepared statements - but also the cache of those on Pgbouncer side is emptied. Running a manual `DEALLOCATE ALL` (or using `pgbouncer=true`) should not cause any problems any more when this is released. Really looking forward to that.

**Comment by MorenoMdz** [comment]
![image](https://github.com/prisma/prisma/assets/29472480/027304f2-5a00-4b53-b653-54591cf586d0)

We have done the upgrade last night and we are seeing solid improvements on all our endpoints that do fetch single records already. 

Interesting note, Aiven upgraded their PG configuration to already include the said fix, making it easier on our side to upgrade 👍 

Thanks @janpio and @JelteF for the hard work here!

**Comment by saadbahir** [comment]
Is it still recommended to use `transaction` mode like suggested in the Prisma [docs](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections/pgbouncer#set-pgbouncer-to-transaction-mode) ?

For reference I removed `pgbouncer=true` alongside `max_prepared_statements > 0` and I get a better latency too

I kept `POOL_MODE=session` but I am not too sure whether that is the right decision

**Comment by JelteF** [comment]
The recommended is:

- max_prepared_statement>0
- pool_mode=transaction
- **remove** pgbouncer=true

**Comment by alessio-zamparelli** [comment]
> The recommended setup is:
> 
>     * max_prepared_statements>0
> 
>     * pool_mode=transaction
> 
>     * **remove** pgbouncer=true

missing "s" on [max_prepared_statements](https://www.pgbouncer.org/config.html#max_prepared_statements)

**Comment by JelteF** [comment]
FYI: I released PgBouncer 1.22.0 just now, which includes the support for DEALLOCATE ALL. So with that release using `pgbouncer=true` won't cause breakage anymore. But `pgbouncer=true` is not needed anymore and since 1.21.0 and it is slower, so it's recommended to remove it from your connection strings. 

So I think all that's needed in this issue is actually updating the Prisma docs to reflect that new recommendation. In particular these two pages at least:
1. https://www.prisma.io/docs/orm/overview/databases/postgresql
2. https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections/pgbouncer

**Comment by MorenoMdz** [comment]
Thank you very much JelteF!

**Comment by sdgluck** [comment]
A note regarding custom schema usage with Prisma and PgBouncer. We use a different schema name than the Prisma default. PgBouncer will complain with the following error when the `?schema=X` parameter is provided in the connection string:

> FATAL: unsupported startup parameter: search_path

A workaround for this is to:

- add the `?schema=X` parameter in the connection string as usual
- configure PgBouncer to ignore the `search_path` parameter
- set the default search path on the database user role

```ini
# in PgBouncer config
ignore_startup_parameters=search_path,extra_float_digits
```

```sql
-- set the search path for database user
ALTER ROLE rolename SET search_path = 'X';
```

My understanding of this workaround is that we are tricking Prisma into thinking it is setting the search path via the query parameter in the connection string, when actually the search path is being set on the database user role. This allows Prisma to derive the schema used by the client.

It would be nice if Prisma accepted the schema name as part of the Prisma schema file, so that the client can be configured to use a custom schema without deriving this from the connection string.

**Comment by janpio** [comment]
> It would be nice if Prisma accepted the schema name as part of the Prisma schema file, so that the client can be configured to use a custom schema without deriving this from the connection string.

The problem would unfortunately persist - as Prisma only uses the `?schema=` to run a `SET search_path TO ...` statement upon connecting to the database - which it would also need to do when it was provided via the Prisma schema.

So I think telling PgBouncer not to react to the `search_path` query, but instead then setting the search path via another way (like setting it for the role) seems like the way to go.

**Comment by biw** [comment]
> FYI: I released PgBouncer 1.22.0 just now, which includes the support for DEALLOCATE ALL. So with that release using `pgbouncer=true` won't cause breakage anymore. But `pgbouncer=true` is not needed anymore and since 1.21.0 and it is slower, so it's recommended to remove it from your connection strings.
> 
> So I think all that's needed in this issue is actually updating the Prisma docs to reflect that new recommendation. In particular these two pages at least:
> 
> 1. [prisma.io/docs/orm/overview/databases/postgresql](https://www.prisma.io/docs/orm/overview/databases/postgresql)
> 2. [prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections/pgbouncer](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections/pgbouncer)

Hey @JelteF, I'm currently working on updating the Prisma docs to match this ticket, so I just wanted to double-check. Is it still recommended to do the above with the release of a newer PgBouncer?
